# Исследование надежности заемщиков


# Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

# Описание данных
* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита


# Импорт используемых библиотек

In [ ]:
import pandas as pd


## Загрузка данных и вывод общей информации

In [ ]:
try:
    data = pd.read_csv('/datasets/data.csv')
except FileNotFoundError:
    print('Кажется, такого файла не существует...')

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`.  Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится для исследования.

**Выведем перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Повторный вывод перечня уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Убедимся, что все пропуски заполнены.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведём их к нижнему регистру.**

In [ ]:
data['education'] = data['education'].str.lower()

**Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследование данных

#### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
import matplotlib.pyplot as plt

children_pivot = data.pivot_table(index = 'children', values = 'debt', aggfunc = ['count', 'mean']) # создание сводной таблицы
children_pivot.columns = ['Количество заемщиков', 'Процент должников'] # изменение названия столбцов для наглядности
children_pivot['Количество должников'] = children_pivot['Количество заемщиков'] * children_pivot['Процент должников'] # подсчёт количества должников

children_pivot['Процент вернувших кредит вовремя'] = (children_pivot['Количество заемщиков'] - children_pivot['Количество должников']) / children_pivot['Количество заемщиков'] # подсчёт процента вернувших кредит вовремя
print(children_pivot)

          Количество заемщиков  Процент должников  Количество должников  \
children                                                                  
0                        14091           0.075438                1063.0   
1                         4808           0.092346                 444.0   
2                         2052           0.094542                 194.0   
3                          330           0.081818                  27.0   
4                           41           0.097561                   4.0   
5                            9           0.000000                   0.0   

          Процент вернувших кредит вовремя  
children                                    
0                                 0.924562  
1                                 0.907654  
2                                 0.905458  
3                                 0.918182  
4                                 0.902439  
5                                 1.000000  


**Вывод:** Зависимость наблюдается, но не сильная. Для семей не имеющих детей можно заметить, что почти нет задолженности по кредиту (из-за стабильного финансового положения, без затрат на детей). Для семей с 1 ребёнком наблюдается больше должников, но всё же основное количество не имеют задолженности по кредиту. Остальные семьи (с 3, 4, 5 детьми) не стоит брать в расчёт, так как их количество очень мало по сравнению с семьями, в которых нет детей или 1 ребёнок.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
mt_pivot = data.pivot_table(index = 'family_status', values = 'debt', aggfunc = ['count', 'sum'])
mt_pivot.columns = ['Общее количество', 'Количество заемщиков']
mt_pivot['Процент'] = mt_pivot['Количество заемщиков'] / mt_pivot['Общее количество']

print(mt_pivot.sort_values(by = 'Процент')) # вывод отсортированный по столбцу "Процент"

                       Общее количество  Количество заемщиков   Процент
family_status                                                          
вдовец / вдова                      951                    63  0.066246
в разводе                          1189                    84  0.070648
женат / замужем                   12261                   927  0.075606
гражданский брак                   4134                   385  0.093130
Не женат / не замужем              2796                   273  0.097639


**Вывод:** Зависимость наблюдается. Люди находящиеся в статусе "не женат / не замужем", а также в статусе "гражданский брак" имеют самую наименьшую вероятность по возврате кредита в срок (причины могут быть совершенно разные, наличие ребёнка, молодой возраст). Люди, находящиеся в статусе "вдовец / вдова" и "в разводе" имеют самую наибольшую вероятность по возврате кредита в срок (скорее из-за отвественности).

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
fin_status_pivot = data.pivot_table(index = 'total_income_category', values = 'debt', aggfunc = ['count', 'sum'])
fin_status_pivot.columns = ['Общее кол-во людей', 'Кол-во заемщиков']
fin_status_pivot['Процент'] = round(fin_status_pivot['Кол-во заемщиков'] / fin_status_pivot['Общее кол-во людей'], 3)

print(fin_status_pivot.sort_values(by = 'Процент'))

                       Общее кол-во людей  Кол-во заемщиков  Процент
total_income_category                                               
D                                     349                21    0.060
B                                    5014               354    0.071
A                                      25                 2    0.080
C                                   15921              1353    0.085
E                                      22                 2    0.091


**Вывод:** Зависимость наблюдается. Люди, имеющие самый низкий заработок, имеют наименьшую вероятность возврата кредита в срок (из-за тяжелого финансового положения). Люди, получающие зарплату в среднем от 30 до 50 тысяч, имеют наибольшую вероятность по возврате кредита в срок (возможно из-за того, что не берут большие кредиты и им не тяжело вернуть кредит вовремя). Люди, получающие от 50 до 200 тысяч, имеют, как и люди получающие наименьший зароботок, имеют низкую вероятность возврата кредита в срок (скорее из-за того, что это семьи, так как составляют наибольшую часть людей, беруших кредит). Люди, получающие от 200 тысяч до 1000000 имеют большую вероятность по возврате кредита в срок (из-за хорошего финансового положения). Люди, получающие от 1000000 имеют среднюю вероятность по возврате кредита в срок, среди всех остальных.

#### Как разные цели кредита влияют на его возврат в срок?

In [ ]:
point_pivot = data.pivot_table(index = 'purpose_category', values = 'debt', aggfunc = ['count', 'sum'])
point_pivot.columns = ['Общее кол-во людей', 'Кол-во заемщиков']
point_pivot['Процент'] = round(point_pivot['Кол-во заемщиков'] / point_pivot['Общее кол-во людей'], 3)

print(point_pivot.sort_values(by = 'Процент'))

                          Общее кол-во людей  Кол-во заемщиков  Процент
purpose_category                                                       
операции с недвижимостью               10751               780    0.073
проведение свадьбы                      2313               183    0.079
операции с автомобилем                  4279               400    0.093
получение образования                   3988               369    0.093


**Вывод:** Люди, берущие кредит с целями "получение образования" и "операции с автомобилем" имеют одинаково маленкуью вероятность возврата кредита в срок вовремя. Люди, берущие кредит по оставшимся целям таким, как "операции с недвижимостью" и "проведение свадьбы", имеют большую вероятность вернуть кредит в срок (скорее всего, это из-за суммы кредита).

#### Возможные причины появления пропусков в исходных данных.

*Ответ:* В первую очередь - человеческий фактор. Далее, возможны какие-то технические неполадки.

#### Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?

*Ответ:* Медиана менее чувствительна к выбросам, чем среднее значение, что делает ее более устойчивой.

### Общий вывод.

В проделанной работе:
1. Предобработали данные
    1. Удалили данные с аномальными значениями
    2. Заполнили пропуски медианными значениями
2. Сделали выводы по изученным данным
    1. Наблюдается не сильная зависимость между количеством детей и возврата кредита в срок. Семьи без детей имеют меньше вероятность по возврате кредита в срок, чем семьи с 1 ребёнком (разница не существенная). Процент должников:
        * Семья без детей - 7,5 % от 14091
        * Семья с 1 ребёнком - 9,2 % от 4808
    2. Наблюдается зависимость между семейным положением и возврата кредита в срок. Люди, находящиеся в статусе "не женат/не замужем" или "гражданский брак" имеют меньше вероятность по возврате кредита в срок, чем люди, находящиееся в статусе "вдовец/вдова". Процент должиков:
        * Вдовец/вдова - 6,6 % от 951
        * Не женат/не замужем - 9,7 % от 2796
        * Гражданский брак - 9,3 % от 4134
    3. Наблюдается зависимость между уровнем дохода и возврата кредита в срок. Люди, находящиеся в категории C имеют меньше вероятноть по возврате кредита в срок, чем люди, находящиеся в категории B. Процент должников:
        * Категория B - 7,1 % от 5014
        * Категория C - 8,5 % от 15921
    4. Наблюдается зависимость между целью кредита и его возврата в срок. Люди с целью кредита "операции с автомобилем" или "получение образования" имеют меньшую вероятность, чем люди с целью кредита "операции с недвижимостью"
        * Операции с автомобилем - 9,3 % от 4279
        * Получение образования - 9,3 % от 3988
        * Операции с недвижимостью - 7,3 % от 10751

        
Приведём пример для того, чтобы убедиться в правильности подведённых итогов.
Мужчина Иванов Иван решил взять кредит. При этом, мы знаем о нём несколько фактов:
* семейное положение - гражданский брак
* уровень дохода - 70 тысяч
* количество детей - 0
* цель кредита - операции с автомобилем

Исходя из этих данных:
1. У людей со статусом гражданский брак - 9,3 % человек не вернувших кредит вовремя
2. У людей с уровнем дохода от 50 до 200 тысяч - 8,5 % человек не вернувших кредит вовремя
3. У людей имеющих 4 детей - 9,7 % человек не вернувших кредит вовремя
4. У людей с целью кредита операции с автомобилем -  9,3 % человек не вернувших кредит вовремя

Объединив все эти данные, с большой вероятностью можем сказать, что Иван не вернёт кредит вовремя.